<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Hands-on Lab: Interactive Visual Analytics with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [60]:
import piplite
await piplite.install(['folium'])
await piplite.install(['pandas'])

In [61]:
import folium
import pandas as pd

In [62]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/v4/DV0101EN-Exercise-Generating-Maps-in-Python.ipynb)


In [63]:
## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [64]:
# Download and read the `spacex_launch_geo.csv`
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [65]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [66]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [67]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [68]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
for site_idx in launch_sites_df.index:
    coords = [launch_sites_df.loc[site_idx,"Lat"],launch_sites_df.loc[site_idx,"Long"]]
    label = launch_sites_df.loc[site_idx,"Launch Site"]
    circle = folium.Circle(coords,radius=1000, color='#d35400', fill=True).add_child(folium.Popup(label))
    # Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
    marker = folium.map.Marker(
        coords,
        # Create an icon as a text label
        icon=DivIcon(
            icon_size=(20,20),
            icon_anchor=(0,0),
            html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % label,
            )
        )
    site_map.add_child(circle)
    site_map.add_child(marker)
site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


All sites are at north of the equator line. Three are located at about 28 degrees N and another at 34.
With regards to longitudes, they are situated in positions such that they are all on the coasts of the USA: the first three are in Florida ant the other near LA, in California.

In [69]:
# Task 2: Mark the success/failed launches for each site on the map

Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [70]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [71]:
marker_cluster = MarkerCluster()

*TODO:* Create a new column in `spacex_df` dataframe called `marker_color` to store the marker colors based on the `class` value


In [72]:
# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
marker_colors = ["green" if elem == 1 else "red" for elem in spacex_df.loc[:,"class"]]
spacex_df.loc[:,"marker_color"] = marker_colors

In [73]:
spacex_df.tail()

Launch Site        Lat      Long  class marker_color
51  CCAFS SLC-40  28.563197 -80.57682      0          red
52  CCAFS SLC-40  28.563197 -80.57682      0          red
53  CCAFS SLC-40  28.563197 -80.57682      0          red
54  CCAFS SLC-40  28.563197 -80.57682      1        green
55  CCAFS SLC-40  28.563197 -80.57682      0          red

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [74]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
    # TODO: Create and add a Marker cluster to the site map
    coords = [spacex_df.loc[index,"Lat"],spacex_df.loc[index,"Long"]]
    label = spacex_df.loc[index,"Launch Site"]
    color = spacex_df.loc[index,"marker_color"]
    icon=folium.Icon(color=color)
    marker = folium.Marker(
                location=coords,
                popup=label,
                icon=icon
             )
    marker_cluster.add_child(marker)
site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


In [75]:
# TASK 3: Calculate the distances between a launch site to its proximities

Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [76]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [77]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [78]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

In [79]:
coastline_lat = 34.63499
coastline_lon = -120.6253
coastline_coords = [coastline_lat, coastline_lon]
site_coords = launch_sites_df.loc[3,:]["Lat"], launch_sites_df.loc[3,:]["Long"]
distance_coastline = calculate_distance(site_coords[0], site_coords[1], coastline_lat, coastline_lon)
distance_coastline

1.3534386358713832

In [80]:
railway_lat = 34.6347
railway_lon = -120.62443
railway_coords = [railway_lat, railway_lon]
distance_railway = calculate_distance(site_coords[0], site_coords[1], railway_lat, railway_lon)
distance_railway

1.2694856460343722

In [81]:
city_lat = 34.639476
city_lon = -120.479388
city_coords = [city_lat, city_lon]
distance_city = calculate_distance(site_coords[0], site_coords[1], city_lat, city_lon)
distance_city

12.044180968663806

In [82]:
hw_lat = 34.63899
hw_lon = -120.45789
hw_coords = [hw_lat, hw_lon]
distance_hw = calculate_distance(site_coords[0], site_coords[1], hw_lat, hw_lon)
distance_hw

14.0057318274451

In [83]:
coastline_lat_e1 = 28.59748
coastline_lon_e1 = -80.5842
coastline_lat_e2 = 28.56352
coastline_lon_e2 = -80.56796
coastline_coords_e1 = [coastline_lat_e1, coastline_lon_e1]
coastline_coords_e2 = [coastline_lat_e2, coastline_lon_e2]
site_coords_e0 = launch_sites_df.loc[0,:]["Lat"], launch_sites_df.loc[0,:]["Long"]
site_coords_e1 = launch_sites_df.loc[1,:]["Lat"], launch_sites_df.loc[1,:]["Long"]
site_coords_e2 = launch_sites_df.loc[2,:]["Lat"], launch_sites_df.loc[2,:]["Long"]

distance_coastline0 = calculate_distance(site_coords_e0[0], site_coords_e0[1], coastline_coords_e2[0], coastline_coords_e2[1])
distance_coastline1 = calculate_distance(site_coords_e1[0], site_coords_e1[1], coastline_coords_e2[0], coastline_coords_e2[1])
distance_coastline2 = calculate_distance(site_coords_e2[0], site_coords_e2[1], coastline_coords_e1[0], coastline_coords_e1[1])

In [84]:
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

In [85]:
railway_lat_e1 = 28.52059
railway_lon_e1 = -80.78753
railway_coords_e1 = [railway_lat_e1, railway_lon_e1]

distance_railway0 = calculate_distance(site_coords_e0[0], site_coords_e0[1], railway_coords_e1[0], railway_coords_e1[1])
distance_railway1 = calculate_distance(site_coords_e1[0], site_coords_e1[1], railway_coords_e1[0], railway_coords_e1[1])
distance_railway2 = calculate_distance(site_coords_e2[0], site_coords_e2[1], railway_coords_e1[0], railway_coords_e1[1])

In [86]:
hw_lat_e1 = 28.55757
hw_lon_e1 = -80.7979
hw_coords_e1 = [hw_lat_e1, hw_lon_e1]

distance_hw0 = calculate_distance(site_coords_e0[0], site_coords_e0[1], hw_coords_e1[0], hw_coords_e1[1])
distance_hw1 = calculate_distance(site_coords_e1[0], site_coords_e1[1], hw_coords_e1[0], hw_coords_e1[1])
distance_hw2 = calculate_distance(site_coords_e2[0], site_coords_e2[1], hw_coords_e1[0], hw_coords_e1[1])

In [87]:
city_lat_e1 = 28.61135
city_lon_e1 = -80.80719
city_coords_e1 = [city_lat_e1, city_lon_e1]

distance_city0 = calculate_distance(site_coords_e0[0], site_coords_e0[1], city_coords_e1[0], city_coords_e1[1])
distance_city1 = calculate_distance(site_coords_e1[0], site_coords_e1[1], city_coords_e1[0], city_coords_e1[1])
distance_city2 = calculate_distance(site_coords_e2[0], site_coords_e2[1], city_coords_e1[0], city_coords_e1[1])

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [109]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site

In [88]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
distances = [distance_coastline, distance_railway, distance_city, distance_hw]
coordonates = [coastline_coords, railway_coords, city_coords, hw_coords]
for distance, coords in zip(distances, coordonates):
    icon=folium.Icon()
    distance_marker = folium.Marker(
       coords,
       icon=icon,
        popup=f"{distance:.2f} km",
       )
    lines=folium.PolyLine(locations=[coords, [launch_sites_df.loc[3,:]["Lat"], launch_sites_df.loc[3,:]["Long"]]])
    site_map.add_child(distance_marker)
    site_map.add_child(lines)

In [89]:
distances_e0 = [distance_coastline0, distance_railway0, distance_city0, distance_hw0]
coordonates_e0 = [coastline_coords_e2, railway_coords_e1, city_coords_e1, hw_coords_e1]
for distance, coords in zip(distances_e0, coordonates_e0):
    icon=folium.Icon()
    distance_marker = folium.Marker(
       coords,
       icon=icon,
        popup=f"{distance:.2f} km",
       )
    lines=folium.PolyLine(locations=[coords, [launch_sites_df.loc[0,:]["Lat"], launch_sites_df.loc[0,:]["Long"]]])
    site_map.add_child(distance_marker)
    site_map.add_child(lines)

In [90]:
distances_e1 = [distance_coastline1, distance_railway1, distance_city1, distance_hw1]
coordonates_e1 = [coastline_coords_e2, railway_coords_e1, city_coords_e1, hw_coords_e1]
for distance, coords in zip(distances_e1, coordonates_e1):
    icon=folium.Icon()
    distance_marker = folium.Marker(
       coords,
       icon=icon,
        popup=f"{distance:.2f} km",
       )
    lines=folium.PolyLine(locations=[coords, [launch_sites_df.loc[1,:]["Lat"], launch_sites_df.loc[1,:]["Long"]]])
    site_map.add_child(distance_marker)
    site_map.add_child(lines)

In [91]:
distances_e2 = [distance_coastline2, distance_railway2, distance_city2, distance_hw2]
coordonates_e2 = [coastline_coords_e1, railway_coords_e1, city_coords_e1, hw_coords_e1]
for distance, coords in zip(distances_e1, coordonates_e1):
    icon=folium.Icon()
    distance_marker = folium.Marker(
       coords,
       icon=icon,
        popup=f"{distance:.2f} km",
       )
    lines=folium.PolyLine(locations=[coords, [launch_sites_df.loc[2,:]["Lat"], launch_sites_df.loc[2,:]["Long"]]])
    site_map.add_child(distance_marker)
    site_map.add_child(lines)

In [92]:
site_map

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.

Answers:
VAFB SLC-4E site on the west coast is quite close, ~1.3 km, to the coastline and railway and 12 to 14 km to nearest town and highway.
It seems that for this location we have a conclusion.

The other locations on the east coast are on islands, but not that isolated. Distances to coastlines are under 1 km and to nearest town and highways over 20 km.

# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/)


<!--## Change Log--!>


<!--| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|--!>


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
